gaeguli/pipeline.c

```diff
@@ -528,12 +612,36 @@ _set_stream_caps (GaeguliPipeline * self, GaeguliVideoResolution resolution,
       break;
   }
 
-  caps = gst_caps_new_simple ("video/x-raw", "width", G_TYPE_INT, width,
-      "height", G_TYPE_INT, height,
-      "framerate", GST_TYPE_FRACTION, framerate, 1, NULL);
-  capsfilter = gst_bin_get_by_name (GST_BIN (self->pipeline), "caps");
+  for (i = 0; supported_formats[i] != NULL; i++) {
+    g_autoptr (GstCaps) subset_caps = NULL;
+
+    caps =
+        gst_caps_new_simple (supported_formats[i], "width", G_TYPE_INT, width,
+        "height", G_TYPE_INT, height, "framerate", GST_TYPE_FRACTION, framerate,
+        1, NULL);
+
+    if (!self->source_device_caps)
+      break;
+
+    subset_caps = gst_caps_intersect (caps, self->source_device_caps);
+    subset_caps = gst_caps_simplify (subset_caps);
+
+    if (subset_caps == NULL)
+      continue;
+
+    if (gst_caps_get_size (subset_caps) > 0)
+      break;
+
+    gst_caps_unref (caps);
+    caps = NULL;
+  }
```

Comment:
Why are you computing intersection of requested caps with the device's? This shouldn't be necessary as the caps you set on the capsfilter don't have to be fixed (i.e. describe exactly one format), but it can be a list of formats you support and you leave up to the elements to negotiate concrete caps on startup. You're basically doing part of GStreamer's work with this code.

IMO this for loop can be as simple as:
```
caps = gst_caps_new_empty ();

for (i = 0; supported_formats[i] != NULL; i++) {
  gst_caps_append (caps,
      gst_caps_new_simple (supported_formats[i],
          "width", G_TYPE_INT, width,
          "height", G_TYPE_INT, height,
          "framerate", GST_TYPE_FRACTION, framerate, 1,
          NULL));
}
```
... and then you set those caps on the capsfilter, telling "Give me video in this resolution and framerate, I don't care if it's raw or jpeg, I can deal with both".

Consequently, you can remove anything that maintains `self->source_device_caps`, first of all `_set_source_device_caps()`.

Reply:
Well, "I don't care if it's raw or jpeg" is a little lie - the order of `GstStructure`s in the caps determines the format preference, so if `video/raw` comes first, it should be used whenever the peer's caps permit.

Reply:
Thanks for your review. I see your point. I agree that you approach is far more simple.
Thanks.